tuf/ngclient/updater.py

```diff
@@ -290,8 +291,17 @@ def _load_local_metadata(self, rolename: str) -> bytes:
             return f.read()
 
     def _persist_metadata(self, rolename: str, data: bytes):
-        with open(os.path.join(self._dir, f"{rolename}.json"), "wb") as f:
-            f.write(data)
+        """Acts as an atomic write operation to make sure
+        that if the process of writing is halted, at least the
+        original data is left intact.
+        """
+
+        original_filename = os.path.join(self._dir, f"{rolename}.json")
+        with tempfile.NamedTemporaryFile(
+            dir=self._dir, delete=False
+        ) as temp_file:
+            temp_file.write(data)
+        os.replace(temp_file.name, original_filename)
```

Comment:
I'm still a little concerned about not proactively deleting temporary files. One can imagine a very long-lived process clogging up disk space.

Reply:
Well then, should I flip the `delete=False` statement? @jku Please comment as well.

Reply:
So just to document what happens here: 
* if `write` and `os.replace` both succeed (as we expect), there will be no temporary files on disk
* we cannot set `delete=True` because you can't do both that and os.replace -- that does not make sense
* Not being able to write metadata to disk is, at least currently, a fatal error: Updater will not continue

I'm not objecting to handling write and os.replace errors as well but I think doing so makes the function significantly more complex and hard to test, something like this:

```
try:
    # all current code  (tempfile, write, os.replace) here
except OSError as e:
    # before raising, delete tempfile if it exists
    if tempfile:
        try:
            os.remove(tempfile.name)
        except FileNotFoundError:
            pass
    raise e
```

Testing this kind of code (in a way that actually checks that I got it correct) can be a bit of a pain.... Now that I think about it, making `original_filename` read-only would test at least one case, maybe that would be enough. Hmmm.